File: 2016/10_BalanceBots/bots.ts
// ======================================================================
// Balance Bots
//   Advent of Code 2016 Day 10 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           b o t s . t s
//
// A solver for the Advent of Code 2016 Day 10 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                                  types
// ----------------------------------------------------------------------

interface GoesType {
  value: number;
  bot: number;
}

interface WhichType {
  which: string;
  num: number;
}

interface GivesDataType {
  low: WhichType;
  high: WhichType;
}

interface CompareType {
  low: number;
  high: number;
}

type HasType = number[];

type BotType = Record<number, HasType>;

type OutType = Record<number, HasType>;

type GivesType = Record<number, GivesDataType>;

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

// ======================================================================
//                                                                   Bots
// ======================================================================

export class Bots {
  // Object for Balance Bots
  text: string[];

  part2: boolean;

  bots: BotType;

  outs: OutType;

  goes: GoesType[];

  gives: GivesType;

  hasTwo: HasType;

  constructor(text: string[], part2 = false) {
    // Create a Bot object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.bots = {};
    this.outs = {};
    this.goes = [];
    this.gives = {};
    this.hasTwo = [];

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      this.processText();
    }
  }

  processText(): void {
    // 1. Start with nothing
    this.goes = [];
    this.gives = {};
    // 2. Loop for all lines of text
    this.text.forEach((line) => {
      // 3. Divide the line into parts
      const parts = line.split(' ');
      // 4. Store goes and gives instructions
      //    value 5 goes to bot 2
      //    bot 2 gives low to bot 1 and high to bot 0
      if (parts[0] === 'value') {
        this.goes.push({ value: parseInt(parts[1], 10), bot: parseInt(parts[5], 10) });
      } else {
        this.gives[parseInt(parts[1], 10)] = {
          low: { which: parts[5], num: parseInt(parts[6], 10) },
          high: { which: parts[10], num: parseInt(parts[11], 10) },
        };
      }
    });
  }

  initialDistribution(): void {
    // 1. Reset the values
    this.bots = {};
    this.outs = {};
    this.hasTwo = [];
    // 2. Loop for all the initial distributions
    this.goes.forEach((goes) => {
      // 3. Add the value to the selected bot
      this.giveBot(goes.bot, goes.value);
    });
  }

  giveBot(bot: number, value: number): void {
    if (bot in this.bots) {
      this.bots[bot].push(value);
      if (this.bots[bot].length === 2) {
        this.hasTwo.push(bot);
      }
    } else {
      this.bots[bot] = [value];
    }
  }

  giveOut(out: number, value: number): void {
    if (out in this.outs) {
      this.outs[out].push(value);
    } else {
      this.outs[out] = [value];
    }
  }

  giveToBotOrOut(where: WhichType, value: number): void {
    if (where.which === 'output') {
      this.giveOut(where.num, value);
    } else {
      this.giveBot(where.num, value);
    }
  }

  distribute(bot: number, compare: CompareType): boolean {
    // 0. Precondition
    if (!(bot in this.bots) || !(bot in this.gives) || this.bots[bot].length !== 2) console.log(`Bad distribution bot=${bot}`);
    // 1. Get high and low values
    const high = Math.max(...this.bots[bot]);
    const low = Math.min(...this.bots[bot]);
    // 2. Distribute the high and low values
    this.giveToBotOrOut(this.gives[bot].high, high);
    this.giveToBotOrOut(this.gives[bot].low, low);
    // 3. Now the bot has none
    this.bots[bot] = [];
    // 4. Is this the comparison we were looking for?
    let result = false;
    if (high === compare.high && low === compare.low) result = true;
    return result;
  }

  whoCompares(value1: number, value2: number, verbose = false, limit = 0): number {
    if (verbose) console.log(`whoCompares: (${value1}, ${value2}) ${limit}`);
    // 1. Assume the worst
    let result = NaN;
    // 2. Set the desired value
    const compare: CompareType = {
      high: Math.max(value1, value2),
      low: Math.min(value1, value2),
    };
    // 3. Distribute initial values
    this.initialDistribution();
    // 3. Loop there is nothing to do
    while (this.hasTwo.length > 0) {
      // 4. Distribute from the first bot on the list
      const bot = this.hasTwo.shift() || 0;
      if (this.distribute(bot, compare)) {
        result = bot;
      }
    }
    // 5. Return which bot did the comparison (if any)
    return result;
  }

  multiplyOutputs(verbose = false, limit = 0): number {
    if (verbose) console.log(`multipleOutputs: ${limit}`);
    // 1. Assume the worst
    let result = NaN;
    // 2. Run the bots
    this.whoCompares(99, 99);
    // 3. Check the outputs
    if (0 in this.outs && 1 in this.outs && 2 in this.outs
      && this.outs[0].length === 1 && this.outs[1].length === 1 && this.outs[2].length === 1) {
      result = this.outs[0][0] * this.outs[1][0] * this.outs[2][0];
    }
    // 4. Return the product or NaN
    return result;
  }

  solution(verbose = false, limit = 0): number {
    if (verbose) console.log(`solution: ${limit}`);
    if (this.part2) {
      return this.multiplyOutputs(verbose, limit);
    }
    return this.whoCompares(61, 17, verbose, limit);
  }

  partOne(verbose = false, limit = 0): number {
    // Returns the solution for part one

    return this.solution(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): number {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.solution(verbose, limit);
  }
}

// ======================================================================
// end                          b o t s . t s                         end
// ======================================================================

File: 2016/10_BalanceBots/bots.py
# ======================================================================
# Balance Bots
#   Advent of Code 2016 Day 10 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         b o t s . p y
# ======================================================================
"A solver for the Advent of Code 2016 Day 10 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import bot

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
PART_ONE_VALUES = (61, 17)

# ======================================================================
#                                                                  Bots
# ======================================================================


class Bots(object):   # pylint: disable=R0902, R0205
    "Object for Balance Bots"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.bots = {}
        self.outputs = {}
        self.twos = set()

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            for line in self.text:
                self._process_line(line)

    def _process_line(self, line):
        "Assign values from text"

        assert line is not None and len(line) > 0

        # 1. Get bot number
        number = bot.Bot.get_bot_number(line)
        if number is None:
            print("*** Unable to determine bot number from text: %s" % line)
            return

        # 2. Is this a new bot?
        if number not in self.bots:

            # 3. Add the new bot
            self.bots[number] = bot.Bot(line, self.part2)

        else:
            # 4. Update an existing bot
            self.bots[number].add_text(line)

        # 5. Remember if the bot has two chips
        if self.bots[number].has_two():
            self.twos.add(number)

    def zooming_around(self, values=PART_ONE_VALUES):
        "Have the robots do thier thing"

        # 1. Loop until there is an answer or nothing to do
        while len(self.twos) > 0:

            # 2. Get the number of a robot with somthing to do
            number = self.twos.pop()
            robot = self.bots[number]

            # 3. Check if this is the solution to part one
            if self.part2 is False and robot.am_i_the_one(values):
                return number

            # 4. Get the recipients of the bot's largess
            low_who, low_value, high_who, high_value = robot.gives()

            # 5. Distribute the chips
            if low_who in self.bots:
                if self.bots[low_who].receive(low_value):
                    self.twos.add(low_who)
            else:
                self.outputs[low_who - bot.OUTPUT_BASE] = low_value

            if high_who in self.bots:
                if self.bots[high_who].receive(high_value):
                    self.twos.add(high_who)
            else:
                self.outputs[high_who - bot.OUTPUT_BASE] = high_value

            # 6. Check for the solution to part2
            if 0 in self.outputs and 1 in self.outputs and 2 in self.outputs:
                return self.outputs[0] * self.outputs[1] * self.outputs[2]

        # 7. Return Failure
        return None

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.zooming_around()

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.zooming_around()


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                      b o t s . p y                     end
# ======================================================================
